hbase-server/src/main/java/org/apache/hadoop/hbase/LocalHBaseCluster.java

```diff
@@ -183,6 +177,47 @@ public LocalHBaseCluster(final Configuration conf, final int noMasters,
     }
   }
 
+  /**
+   * Create new LocalHBaseCluster using pre-defined master rpc ports. Every other port is picked
+   * randomly. This also populates the master addresses in the base and region server configs.
+   * @param conf Base configuration to use for the cluster.
+   * @param noMasters Number of masters.
+   * @param noRegionServers Number of region servers.
+   * @param masterClass Impl of master class
+   * @param regionServerClass Impl of RS class
+   * @param masterPorts Array of ports, 2 per master (RPC/INFO) to use.
```

Comment:
What's the use-case for specifying the explicit ports to use? This design has a code-smell to me -- a test should grab whatever ports are available on the system and provide an interface for handing that information off to clients, not the other way around.

Why just offer the master ports as random? If this is something we want to support in tests, seems like we need a `LocalHBaseClusterBuilder` that lets me build up a configuration, adding port pairs for master instances and region server instances (and possible zookeeper, data node, name node, &c).

Reply:
I think I over thought about this. I've undone all of these changes and made the patch quite simple. LMK what you think.

To give you some context, the reason I did this change is to have the master end points (comma separated host:port values) populated in the region server configs so that they can use the newly created master based registry to create the internal connections. The issue though is that since the masters were using random ports chosen at runtime, we cannot pre-create a config object before spinning up the cluster.

But after spending some more time on it, it looks like we can actually do that easily. The next patch just waits for the masters to be created -> figures out the ports -> updates the conf -> spins up the region servers with that conf.